### PR TITLE
Migrate to transformers >= 5

### PR DESCRIPTION
**Pull request by sivanravidos**
## What this PR does

Migrates the codebase to `transformers` v5. The dependency pin moves from
`transformers>=4.54.0,<5.0.0` to `transformers>=5,<6`, and the model, tokenizer,
and checkpoint-loading code is updated for the v5 API. The test suite passes on
transformers 5.

> [!WARNING]
> **transformers < 5 is no longer supported.** After this PR the package requires
> `transformers>=5,<6`. Existing environments must upgrade. Checkpoints trained
> under transformers 4 still load — they are migrated automatically on load (see below).

## Areas changed

**Models** (`scbert`, `scmodernbert`, `scperformer`, `scnystromformer`, `llama`)
- Adapt to v5's weight-initialization flow so pretrained embeddings are not
  re-randomized when `from_pretrained` re-runs init.
- Restore config attribute defaults that v5 dropped from `PretrainedConfig`, and
  accept v5's new `tie_weights` signature.
- Remove the head-mask / head-pruning machinery that v5 no longer provides and
  this codebase never used.
- Compute default RoPE inline for ScModernBert (v5 removed the `'default'` entry
  from its RoPE init registry).

**Tokenizer** (`multifield_tokenizer.py` and friends)
- Load each sub-tokenizer's `tokenizer.json` directly rather than through
  `BertTokenizerFast`, which v5 aliases to the WordPiece `BertTokenizer` and would
  silently mangle the WordLevel gene vocabs and BPE DNA vocabs.
- Re-apply the case-sensitive normalizer and restore special tokens / model input
  names that the direct load would otherwise drop.

**Checkpoint loading** (`training/modules/base.py`)
- Merge model/trainer config from the checkpoint with the run config.
- Automatically migrate older (MLM / SeqCls / SeqLabel) checkpoint layouts to the
  current multitask format on load, so transformers-4 checkpoints keep working.

**anndata compatibility**
- Replace the removed `AnnData.concatenate` with `anndata.concat` (the newer
  anndata pulled in alongside transformers 5 dropped the old method).

**Tests & CI**
- Added tests for the model init/config behavior, tokenizer loading, and checkpoint
  migration.
- Free unused runner disk space before the CLI checkpoint tests (the larger v5
  install left too little room for checkpoint saves).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -27,7 +27,7 @@ dependencies = [
   "torch>=2.0.0",
   "rdata",
   "pytorch-lightning>=2.0.0",
-  "transformers>=4.54.0,<5.0.0",
+  "transformers>=5,<6",
```
Comment (sivanravidos) [inline pyproject.toml:30]: so we give up support transformers <5 ?

File: bmfm_targets/tokenization/multifield_tokenizer.py [hunk 2]
```diff
@@ -281,19 +283,80 @@ def load_subtokenizer(self, field_name: str, field_to_tokenizer_map=None):
...
+            _SPECIAL_TOKEN_KEYS = {
+                "pad_token",
+                "cls_token",
+                "sep_token",
+                "unk_token",
+                "mask_token",
+                "bos_token",
+                "eos_token",
+            }
+            for candidate in ["special_tokens_map.json", "tokenizer_config.json"]:
```
Comment (sivanravidos) [inline bmfm_targets/tokenization/multifield_tokenizer.py]: This reads only the first file found, if both exist the second will be ignored

File: bmfm_targets/models/predictive/scnystromformer/modeling_scnystromformer.py [hunk 3]
```diff
@@ -386,16 +358,23 @@ def _init_weights(self, module):
         if isinstance(module, nn.Linear | nn.Conv2d):
             # Slightly different from the TF version which uses truncated_normal for initialization
             # cf https://github.com/pytorch/pytorch/pull/5617
-            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            # transformers v5 wraps initialize_weights with guard_torch_init_functions, so
```
Comment (sivanravidos) [inline bmfm_targets/models/predictive/scnystromformer/modeling_scnystromformer.py:360]: this is added here and in `modeling_scperformer.py`. maybe put it in a shared place in case it changes again in the future?

File: bmfm_targets/models/predictive/attentions.py [hunk 4]
```diff
@@ -175,41 +167,10 @@ def __init__(self, config: PretrainedConfig, self_attention):
         self.output = SCSelfOutput(config)
         self.pruned_heads: set[int] = set()
```
Comment (sivanravidos) [inline bmfm_targets/models/predictive/attentions.py]: looks like `pruned_heads` here and in the other models is also not used anymore

**Review by sivanravidos — COMMENTED** [review]
thanks for the fixes!!
so we give up transformers<5?

I did a claude assisted review:
- The PR text is not up to date anymore (it's  from my initial changes)
- Claude and I found several minor issues, comments inline

- I did a manual regression check, I ran end-to-end zero-shot inference (inference.py) with wced checkpoint on 20 cells from zheng68k comparing main branch + transformers==4.57.6  to this PR  branch (with transformers==5.13.0) and compared the embeddings, its identical.
- I also run vllm tests on gpu and they pass

File: pyproject.toml (hunk 1 above)
Comment (mmdanziger) [reply pyproject.toml:30]: yea i think that's the best, we do not have bandwidth to maintain two branches.

File: bmfm_targets/models/predictive/attentions.py (hunk 4 above)
Comment (mmdanziger) [reply bmfm_targets/models/predictive/attentions.py]: removed

File: bmfm_targets/tokenization/multifield_tokenizer.py (hunk 2 above)
Comment (mmdanziger) [reply bmfm_targets/tokenization/multifield_tokenizer.py]: fixed with test

File: bmfm_targets/models/predictive/scnystromformer/modeling_scnystromformer.py (hunk 3 above)
Comment (mmdanziger) [reply bmfm_targets/models/predictive/scnystromformer/modeling_scnystromformer.py:360]: let's just leave it, it was like that before and doesn't hurt, may help future agents/devs because it's kinda tricky

**Review by mmdanziger — APPROVED** [review]
